tests/test_kernel.c

```diff
@@ -0,0 +1,444 @@
+	kmbox_send(mbox, &msg4);
+	uint8_t exp_after_4[] = { 1,2,3,4,0 };
+	ASSERT_EQUAL_MEM(msgread, exp_after_4, sizeof(exp_after_4), "invalid order of messages");
+
+	// This will block, and at this point the reader thread will be scheduled.
+	// As soon as it reads one message off the mbox, it will be switched away
+	// without having time to process it.
+	kmbox_send(mbox, &msg5);
+	kmbox_send(mbox, &msg6);
+	kmbox_send(mbox, &msg7);
+	kmbox_send(mbox, &msg8);
+
+	uint8_t exp_final[] = { 1,2,3,4,5,6,7,8,0 };
+	ASSERT_EQUAL_MEM(msgread, exp_final, sizeof(exp_final), "invalid order of messages");
+}
+
+void test_kernel_event_1(TestContext *ctx) {
+	timer_init();
+	DEFER(timer_close());
+
+	kernel_init();
+	DEFER(kernel_close());
+
+	set_SP_interrupt(true);
+	DEFER(set_SP_interrupt(false));
+
+	bool called = false;
+
+	void wait_thread(void *arg)
+	{
+		kthread_wait_event(&KEVENT_IRQ_SP);
+		called = true;
+	}
+
+	void rsp_thread(void* arg)
+	{
+		uint32_t __attribute__((aligned(8))) miniucode[] = { 0x0000000d }; // BREAK opcode
+		
+		data_cache_hit_writeback_invalidate(miniucode, sizeof(miniucode));
+		load_ucode(miniucode, sizeof(miniucode));
+		run_ucode();
+	}
+
+	kthread_set_pri(NULL, 1);
+	kthread_new("wait_thread", 2048, 2, wait_thread, NULL);
+	kthread_new("rsp_thread", 2048, 2, rsp_thread, NULL);
+
+	kthread_sleep(TICKS_FROM_MS(5));
+	ASSERT(called, "event not triggered");
+}
+
+void test_kernel_event_2(TestContext *ctx) {
+	timer_init();
+	DEFER(timer_close());
+
+	kernel_init();
+	DEFER(kernel_close());
+
+	set_SP_interrupt(true);
+	DEFER(set_SP_interrupt(false));
+
+	uint8_t called[16] = {0};
+	int called_idx = 0;
+
+	kmbox_t *m1 = kmbox_new_stack(1);
+	kmbox_t *m2 = kmbox_new_stack(1);
+
+	void read_kthread_1(void* arg)
+	{
+		while (kmbox_recv(m1) == &KEVENT_IRQ_SP)
+			called[called_idx++] = 1;
+	}
+
+	void read_kthread_2(void* arg)
+	{
+		while (kmbox_recv(m2) == &KEVENT_IRQ_SP)
+			called[called_idx++] = 2;
+	}
+
+	void rsp_thread(void* arg)
+	{
+		uint32_t __attribute__((aligned(8))) miniucode[] = { 0x0000000d }; // BREAK opcode
+		
+		data_cache_hit_writeback_invalidate(miniucode, sizeof(miniucode));
+		load_ucode(miniucode, sizeof(miniucode));
+		run_ucode();
+	}
+
+	kmbox_attach_event(m1, &KEVENT_IRQ_SP);
+	DEFER(kmbox_detach_event(m1, &KEVENT_IRQ_SP));
+
+	kmbox_attach_event(m2, &KEVENT_IRQ_SP);
+	DEFER(kmbox_detach_event(m2, &KEVENT_IRQ_SP));
+
+	kthread_set_pri(NULL, 5);
+	kthread_new("rsp_thread", 2048, 2, rsp_thread, NULL);
+
+	kthread_t *rth1 = kthread_new("read_kthread_1", 2048, 3, read_kthread_1, NULL);
+	DEFER(kthread_kill(rth1));
+
+	kthread_t *rth2 = kthread_new("read_kthread_2", 2048, 4, read_kthread_2, NULL);
+	DEFER(kthread_kill(rth2));
+
+	// Go to sleep. Now the RSP will run, trigger the two reader threads, and
+	// they will register themselves in the called array.
+	kthread_sleep(TICKS_FROM_MS(5));
+
+	// Make sure the order is correct (first the higher priority thread).
+	uint8_t exp1[] = { 2, 1, 0 };
```

Comment:
I expected `{ 2, 0 }` because of;
> Once a thread receives a mail from a mailbox, the mail is removed and other threads will not see it anymore

Reply:
there are two mailbox here `m1` and `m2`, and both of them are attached to the same event. Events are broadcast, so all mboxes attached to an event receive it. This is by design to allow multiple threads to wait for the same interrupt.